Make Core Controller application socket connections non-blocking

CamJN opened the pull request:
`ApplicationPool::Socket::checkoutConnection()` initiates blocking connects, which could block the event loop if the application process's socket backlog is full.

CamJN committed:
make connect non blocking

CamJN committed:
setup for handling ready state

FooBarWidget commented on src/agent/Core/ApplicationPool/Socket.h:
```diff
@@ -88,10 +90,12 @@ class Socket {
+		NConnect_State state(address, __FILE__, __LINE__);
+		connection.ready = state.connectToServer();
 		connection.fail = true;
+		connection.fd = state.getFd();
```
This fd will become invalid once this scope is exited. The state structure has a FileDescriptor RAII object which closes the underlying fd as soon as it goes out of scope.

FooBarWidget committed:
Fix NConnect_State file descriptor becoming invalid when leaving connect() scope

FooBarWidget commented on src/agent/Core/ApplicationPool/Session.h:
```diff
@@ -185,22 +187,47 @@ class Session: public AbstractSession {
+		ScopeGuard g(boost::bind(&Session::callOnInitiateFailure, this));
+		Connection connection = socket->checkoutConnection();
+		if (!connection.ready) {
+			new oxt::thread(&waitConnection, connection, blocking, boost::bind(&Session::continueWithConnection, this));
```
Should allow integration with an event loop instead of spawning a thread.

CamJN replied:
`Controller∷initiateSession` passes the fd to `req→appSink.reinitialize` and `req→appSource.reinitialize` which use evented io, so I think it's ok now?

FooBarWidget replied:
It only uses evented I/O to read from/write to the fd. It doesn't wait until the connect finishes. You can't read from/write to the fd until it's actually connected.

FooBarWidget commented on src/agent/Core/ApplicationPool/Socket.h:
```diff
@@ -52,12 +52,14 @@ struct Connection {
 	bool wantKeepAlive: 1;
 	bool fail: 1;
 	bool blocking: 1;
+	bool ready: 1;
```
I think in 'Socket' we can drop support for blocking connects.

There are only two places where we call Session::initiate(): in the Core Controller (CheckoutSession.cpp, needs non-blocking connect) and in PoolTest.cpp (needs blocking connect).

We can modify Session::initiate(bool blocking) as follows:
- Call Socket->checkoutSession() which is always presumed to make a non-blocking connect.
- If `blocking` is true, then use poll() to perform a blocking wait on the socket to become ready. Then set the fd back to blocking mode.

This adds more code, but reduces the number of different code paths that we have to deal with (no entirely separate blocking and non-blocking code paths), and thus should reduce bugs. In the `blocking=true` case (only used by tests) we don't care about performance at all so the extra calls are fine.

FooBarWidget commented on src/agent/Core/ApplicationPool/Session.h:
```diff
@@ -185,22 +187,47 @@ class Session: public AbstractSession {
+		struct timeval timeout;
+		timeout.tv_sec = 0;
+        timeout.tv_usec = 10000;
+		if (0 != syscalls::select(largestFd + 1, &fds, &fds, NULL, &timeout)) {
```
Use poll()

FooBarWidget commented:
A refresher of the different call paths:

session->initiate() called from:
- Core Controller (CheckoutSession.cpp)
- PoolTest.cpp

What does session->initiate(blocking) do:
- calls socket->checkoutConnection()
    - takes a connection from the pool if available
    - otherwise: creates a new connection
- sets session->fd

CamJN committed:
make non blocking case branchless

CamJN commented:
The tests are hanging now. @FooBarWidget any thoughts on what I can do to fix it? I wait on the connection to be ready and then make it blocking in tests.

FooBarWidget commented:
Your PR isn't ready so don't worry about test failures until it is ready. You haven't begun modifying the controller yet so it's no surprise that tests for the controller fail. The ApplicationPool tests pass and that's what matters right now. For now, run only ApplicationPool tests locally to verify that at least that part is ok.

FooBarWidget commented on src/agent/Core/ApplicationPool/Socket.h:
```diff
@@ -69,6 +69,18 @@ struct Connection {
+		pfd.fd = fd;
+		pfd.events = POLLIN | POLLOUT;
+		if (-1 == syscalls::poll(&pfd, 1, timeout)) {
+			throw SystemException("poll() failed", errno);
```
When throwing a SystemException, always first assign errno to a temporary variable. Otherwise the creation of the std::string for the exception might overwrite errno.

FooBarWidget commented on src/agent/Core/ApplicationPool/Session.h:
```diff
@@ -186,19 +186,21 @@ class Session: public AbstractSession {
+		if (blocking) {
+			// for testing purposes
+			if (!connection.ready) { connection.wait(); }
+			setBlocking(connection.fd);
```
If wait or setBlocking fails here then the connection won't be cleaned up since g2 is already cleared.

But now that I think about it: since we *only* care about blocking I/O in the tests, I think it's better to move the wait to the tests. Just don't have blocking I/O support in ApplicationPool at all, and have only the tests deal with that.

FooBarWidget commented on src/agent/Core/ApplicationPool/Session.h:
```diff
@@ -186,19 +186,21 @@ class Session: public AbstractSession {
-			connection.blocking = false;
-		}
+		FdGuard g2(connection.fd, NULL, 0);
+		setNonBlocking(connection.fd);
```
No need to set non-blocking here, it's already non-blocking by default.

CamJN committed:
preserve errno for exception

CamJN committed:
remove blocking code from non-test session code

CamJN commented:
> You haven't begun modifying the controller yet 

Controller∷initiateSession passes the fd to req→appSink.reinitialize and req→appSource.reinitialize which use evented io, so what changes are needed in the controller?

FooBarWidget commented:
> > You haven't begun modifying the controller yet
> 
> Controller∷initiateSession passes the fd to req→appSink.reinitialize and req→appSource.reinitialize which use evented io, so what changes are needed in the controller?

In the controller you need to wait until the non-blocking connect is actually finished connecting. You can't read from/write to the fd until the connect operation is done.

Also be sure to check whether the non-blocking connect failed, by querying `getsockopt(SO_ERROR)`.

FooBarWidget commented on test/cxx/Core/ApplicationPool/PoolTest.cpp:
```diff
@@ -190,7 +190,13 @@ namespace tut {
+			if (!ready) {
+				Connection connection;
+				connection.fd = currentSession->fd();
+				connection.wait();
```
We can move the entire wait() method to PoolTest since it won't be used outside the tests anyway.

FooBarWidget commented on src/agent/Core/ApplicationPool/Socket.h:
```diff
@@ -69,6 +69,19 @@ struct Connection {
+		TRACE_POINT();
+		pollfd pfd;
+		pfd.fd = fd;
+		pfd.events = POLLIN | POLLOUT;
```
You're only supposed to poll for writability, not readability.

But actually you can use IOUtils.h `waitUntilWritable()`.

After the wait is done, you're also supposed to check whether the non-blocking connect failed with an error, by querying `getsockopt(SO_ERROR)`.

From the connect(2) man page:

> EINPROGRESS
>              The socket is nonblocking and the connection cannot be
>              completed immediately.  (UNIX domain sockets failed with
 >             EAGAIN instead.)  It is possible to [select(2)](https://man7.org/linux/man-pages/man2/select.2.html) or [poll(2)](https://man7.org/linux/man-pages/man2/poll.2.html)
 >             for completion by selecting the socket for writing.  After
 >             [select(2)](https://man7.org/linux/man-pages/man2/select.2.html) indicates writability, use [getsockopt(2)](https://man7.org/linux/man-pages/man2/getsockopt.2.html) to read
>              the SO_ERROR option at level SOL_SOCKET to determine
>              whether connect() completed successfully (SO_ERROR is zero)
>              or unsuccessfully (SO_ERROR is one of the usual error codes
>              listed here, explaining the reason for the failure).

FooBarWidget commented on src/agent/Core/ApplicationPool/Socket.h:
```diff
@@ -69,6 +69,19 @@ struct Connection {
+		pollfd pfd;
+		pfd.fd = fd;
+		pfd.events = POLLIN | POLLOUT;
+		if (-1 == syscalls::poll(&pfd, 1, timeout)) {
```
You should handle timeouts. It returns 0 on timeout.

But actually you can use IOUtils.h `waitUntilWritable()`, which already takes care of this.

FooBarWidget committed:
Improve docs

[ci skip]

CamJN committed:
simplify waiting for socket in test

CamJN committed:
fix unsafe use of errno

CamJN committed:
wait for connect in controller

CamJN commented:
The controller test is still hanging:
```
  * frame #0: 0x00007ff808d04aaa libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007ff808d437a8 libsystem_pthread.dylib`_pthread_cond_wait + 1193
    frame #2: 0x000000010f8289fb main`boost::condition_variable::wait(boost::unique_lock<boost::mutex>&) [inlined] boost::posix::pthread_cond_wait(c=0x00007f8789805d28, m=0x00007f8789805ce8) at pthread_helpers.hpp:112:17
    frame #3: 0x000000010f8289ee main`boost::condition_variable::wait(this=0x00007f8789805ce8, m=0x00007ff7b073ac70) at condition_variable.hpp:79:19
    frame #4: 0x000000010f94c84e main`Passenger::SafeLibev::runSync(this=0x00007f8789805c70, callback=0x00007ff7b073ad58) at SafeLibev.h:245:9
    frame #5: 0x000000010f94c3bb main`tut::Core_ControllerTest::~Core_ControllerTest(this=0x00007f878a00a000) at ControllerTest.cpp:109:14
```

Despite now waiting on the socket.

FooBarWidget commented on test/cxx/Core/ApplicationPool/PoolTest.cpp:
```diff
@@ -190,7 +191,16 @@ namespace tut {
-			currentSession->initiate();
+			if (!currentSession->initiate()) {
+				unsigned long long timeout = 10000;
+				while (timeout > 0 && !waitUntilWritable(currentSession->fd(), &timeout)) { }
```
A while loop is not appropriate here. waitUntilWritable works just like poll. If it returns false then that just means the full timeout expired, and you should fail the test.

CamJN replied:
It looks like it breaks the timeout into 1000 chunks and adjusts the timeout pointer as it is called repeatedly… https://github.com/phusion/passenger/blob/stable-6.0/src/cxx_supportlib/IOTools/IOUtils.cpp#L816 https://github.com/phusion/passenger/blob/stable-6.0/src/cxx_supportlib/IOTools/IOUtils.cpp#L864-L879

FooBarWidget replied:
The timeout is in microseconds, that's why the /1000.

The writeExact() code has a while loop, not because waitUntilWritable breaks the timeout into 1000 chunks, but because a write() system call can return early without having written all requested data. writeExact loops until everything is written.

FooBarWidget commented on test/cxx/Core/ApplicationPool/PoolTest.cpp:
```diff
@@ -190,7 +191,16 @@ namespace tut {
+				unsigned long long timeout = 10000;
+				while (timeout > 0 && !waitUntilWritable(currentSession->fd(), &timeout)) { }
+				ensure_gt("(1)", timeout, 0ULL);
+				int connect_error = 0;
```
Should use camelCase naming.

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -178,7 +179,21 @@ Controller::initiateSession(Client *client, Request *req) {
-		req->session->initiate(false);
+		if (!req->session->initiate()) {
+			unsigned long long timeout = 100000;
+			while (timeout > 0 && !waitUntilWritable(req->session->fd(), &timeout)){}
```
This is a blocking wait. In the Controller you're supposed to use a libev io watcher to use the event loop to wait.

CamJN committed:
use camel case

CamJN committed:
fix test

CamJN committed:
use libev callback when session->initiate not ready

CamJN committed:
move socket check into callback

CamJN commented:
It still hangs on linux and now it crashes on macOS. Ideas?

CamJN committed:
ensure cbs is deleted in all code paths

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -173,12 +174,53 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
 	try {
-		req->session->initiate(false);
+		if (!req->session->initiate()) {
+			ev_io connectedWatcher;
```
You can't allocate this on the stack. It has to stay alive while waiting is in progress. Should be in the Request struct.

FooBarWidget replied:
The request could be aborted while waiting for the async connect. Make sure you stop this watcher when that happens (in deinitializeRequest).

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -173,12 +174,53 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
+			connectedWatcher.data = new ControllerFinishInitiatingSessionCallbackPointers(this, client, req);
+			ev_io_init(&connectedWatcher, onWritable, req->session->fd(), EV_WRITE);
+			ev_io_start(getLoop(), &connectedWatcher);
+			return;
```
There should also be a connect timeout.

There's currently no good config option for that, so we should introduce a new per-request config option "app_connect_timeout". I think a default of 30s is appropriate.

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -173,12 +174,53 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
 	req->sessionCheckoutTry++;
 	try {
-		req->session->initiate(false);
+		if (!req->session->initiate()) {
```
Would be good to have some logging here saying that connect is either in progress or finished immediately.

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -173,12 +174,53 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
+		Request * req = cbs->get<2>();
+        ev_io_stop(that->getLoop(), io);
+
+		int connect_error = 0;
```
Should use camelCase.

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -173,12 +174,53 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
+void
+Controller::onWritable(EV_P_ ev_io *io, int revents) {
+        ControllerFinishInitiatingSessionCallbackPointers *cbs = static_cast<ControllerFinishInitiatingSessionCallbackPointers *>(io->data);
+	    Controller* that = cbs->get<0>();
```
No need to use a tuple. Just pass req as data. Then you can get everything else from the req:

```c++
Request *req = static_cast<Request *>(w->data);
Client *client = static_cast<Client *>(req->client);
Controller *self = static_cast<Controller *>(Controller::getServerFromClient(client));
```

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -178,12 +179,53 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
+			throw SystemException("Cannot check socket status", err);
+		} else if (connect_error != 0) {
+			// connect_error uses the same error codes as errno
+			throw SystemException("Cannot connect socket", connect_error);
```
We're in an event loop callback. So rather than throwing an exception, any errors we counter should either result in:

1. a call to disconnectWithError() (if it's a permanent error), or,
2. retrying the session initiation, like what you see in initiateSession's exception handler.

FooBarWidget committed:
Fix TestSession::initiate(): return non-blocking connection

FooBarWidget commented:
> It still hangs on linux and now it crashes on macOS. Ideas?

It's because ControllerTest uses TestSession instead of a real Session, and TestSession::initiate() still returns a blocking fd, so the event loop was stuck on a blocking read. I fixed that.

The test suite is still broken by other things though.

CamJN committed:
update changelog

[ci skip] [ci:skip]

CamJN committed:
move ev_io to request object

CamJN committed:
remove tuple, use request to get all info

CamJN committed:
stop connectionWatcher if request stopped

CamJN committed:
add logging of async state

CamJN committed:
add timeout code

CamJN committed:
stop connectionWatcherTimeout if request stopped

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -178,36 +179,110 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
+			req->connectedWatcher.data = req;
+			ev_io_init(&req->connectedWatcher, onWritable, req->session->fd(), EV_WRITE);
+			// only one of timeout or ready handler should fire, so only ref once
+			refRequest(req, __FILE__, __LINE__);
```
refRequest is only necessary when you're making a non-libev async operation that can't be cancelled, in order to keep req alive until the completion callback. Since libev async operations can be cancelled, use of refRequest is not necessary.

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -178,36 +179,110 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
 }
 
+void
+Controller::onTimeout(EV_P_ ev_timer *io, int flag) {
```
Should have more specific name that indicates what timed out.

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -178,36 +179,110 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
+}
+
+void
+Controller::onWritable(EV_P_ ev_io *io, int revents) {
```
Should have more specific name.

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -178,36 +179,110 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
+	Client *client = static_cast<Client *>(req->client);
+	Controller *that = static_cast<Controller *>(Controller::getServerFromClient(client));
+    ev_io_stop(that->getLoop(), &req->connectedWatcher);
+	SystemException e("Waiting on socket connect timed out", ETIMEDOUT);
```
You're creating an exception here so this function should have a TRACE_POINT() for better backtraces.

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -178,36 +179,110 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
+		Controller *that = static_cast<Controller *>(Controller::getServerFromClient(client));
+        ev_io_stop(that->getLoop(), io);
+
+		if (revents & EV_WRITE) {
```
You're creating an exception here so this function should have a TRACE_POINT() for better backtraces.

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -178,36 +179,110 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
+			that->finishInitiatingSession(client, req);
+        } else {
+			// something went very wrong
+			int err = errno;
```
This is not an error that can be inspected through errno.

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -178,36 +179,110 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
+		Request *req = static_cast<Request *>(io->data);
+		Client *client = static_cast<Client *>(req->client);
+		Controller *that = static_cast<Controller *>(Controller::getServerFromClient(client));
+        ev_io_stop(that->getLoop(), io);
```
You also need to cancel the timeout.

FooBarWidget commented on src/agent/Core/Controller/Request.h:
```diff
@@ -112,7 +114,10 @@ class Request: public ServerKit::BaseHttpRequest {
 	Request()
 		: BaseHttpRequest()
-		{ }
+		{
```
For consistency it's better to put this code in reinitializeRequest()

FooBarWidget commented on src/agent/Core/Controller/Hooks.cpp:
```diff
@@ -154,6 +154,15 @@ Controller::reinitializeRequest(Client *client, Request *req) {
+		ev_io_stop(that->getLoop(), &req->connectedWatcher);
+	}
+	if (req->connectedWatcherTimout.active) {
+		Controller *that = static_cast<Controller *>(Controller::getServerFromClient(client));
```
You're already in the controller.

FooBarWidget committed:
update changelog

[ci skip]

CamJN committed:
add config option for app connect timeout

CamJN committed:
remove ref counting calls

CamJN committed:
rename callbacks to be more descriptive

CamJN committed:
put request initialization into reinitialize

CamJN committed:
no need to lookup controller from in non static member function

CamJN committed:
add tracepoints

CamJN committed:
stop timer when socket connects

FooBarWidget commented on src/agent/Core/ApplicationPool/Options.h:
```diff
@@ -211,6 +211,11 @@ class Options {
+	/** The maximum amount of time, in milliseconds, that may be spent
+	 * waiting on the process to accept controller's connection on the
+	 * socket. */
+	unsigned int connectTimeout;
```
No need to propagate this to ApplicationPool. We only use this option in the Controller.

FooBarWidget reviewed (commented):
Almost there. And tests still needed.
- Testing async connect flow.
- Testing async connect timeout.
- Testing disconnect while async connect in progress.

FooBarWidget committed:
minor refactoring

FooBarWidget commented on src/agent/Core/Controller/CheckoutSession.cpp:
```diff
@@ -178,29 +179,52 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
+			ev_timer_start(getLoop(), &req->connectedWatcherTimout);
+
+			req->connectedWatcherTimout.data = req;
+			ev_timer_init(&req->connectedWatcherTimout, onSessionSocketConnectTimeout, req->options.connectTimeout / 1000.0, 0);
```
It's not entirely clear to me whether we're handling seconds <-> milliseconds conversion properly. constants.rb is in milliseconds but it seems web server options are not. Need to double check later when you're done removing the ApplicationPool option.

CamJN replied:
To make it clearer I just switched to using seconds throughout the codebase for that option.

CamJN committed:
don’t save connect timeout in pool options

FooBarWidget committed:
refactorings

FooBarWidget committed:
refactor

CamJN committed:
Use seconds throughout connect timeout code

CamJN commented:
@FooBarWidget the cache steps in the CI seem to be pretty flakey. One has failed in each of the last several runs.

FooBarWidget commented:
Looks fine so far. Just need tests.

I'm not entirely happy about the fillPoolOption() call, but the config handling in the Controller is in general just messed up, and it needs a bigger rework, so I'll just leave it like that for now.

Hope we can release this week.

FooBarWidget commented:
> @FooBarWidget the cache steps in the CI seem to be pretty flakey. One has failed in each of the last several runs.

There was a cache corruption which I fixed. Other errors are just normal flakiness, you can retry.

FooBarWidget commented on src/agent/Core/ApplicationPool/TestSession.h:
```diff
@@ -55,6 +56,7 @@ using namespace std;
  */
 class TestSession: public AbstractSession {
 private:
+	typedef boost::tuple<TestSession*, FileDescriptor> AsyncTuple;
```
Better use std::tuple, we want to gradually use more C++11/C++14 std stuff over Boost.

FooBarWidget commented on test/cxx/Core/ControllerTest.cpp:
```diff
@@ -1144,4 +1144,66 @@ namespace tut {
+		set_test_name("Testing async connect flow.");
+		init();
+		mockNextSession();
+		struct ev_async& async = testSession.setupAsync(bg.libev_loop);
```
Your strategy with setting up an ev_sync boils down to delaying the accept() call until a moment you choose. But there's really no need to delay the accept call. If you just want to test the async connect flow, then you *only* have to make testSession->initiate() return false, you don't need anything else.

CamJN replied:
I need to delay the accept call so that I can make it take until the connect timeout has run out or the client disconnect has happened.

FooBarWidget commented on test/cxx/Core/ControllerTest.cpp:
```diff
@@ -1144,4 +1144,66 @@ namespace tut {
+			"Host: localhost\r\n"
+			"Connection: close\r\n"
+			"\r\n");
+	    EVENTUALLY(5, { result = async.data != NULL; });
```
This is not thread-safe. async.data is mutated in a different thread. You need a different, synchronized, tactic.

FooBarWidget commented on test/cxx/Core/ControllerTest.cpp:
```diff
@@ -1144,4 +1144,66 @@ namespace tut {
+		waitUntilSessionInitiated();
+    }
+
+	TEST_METHOD(61) {
```
This test strategy does not work. You are making TestSession.initiate() never call accept(). But the connect() still finishes immediately (i.e. the ev_io watcher on the socket still gets an EV_WRITE event soon, before the timeout fires).

CamJN committed:
make connection wait in tests only

CamJN committed:
pass on ready state to caller of session initiate

CamJN committed:
test attempt 2

FooBarWidget commented on test/cxx/Core/ControllerTest.cpp:
```diff
@@ -56,6 +57,9 @@ namespace tut {
 				mockException.reset();
 			}
+			virtual int ioConditions() override {
+				return EV_NONE;
```
You are *always* returning EV_NONE here but you should only do that in the tests where you want to mock this.

CamJN committed:
put missing refRequest back in handleSessionInitiationError

CamJN committed:
update tests

CamJN committed:
disconnect with gateway timeout when cannot connect to app

FooBarWidget committed:
update headers

[ci skip]

FooBarWidget committed:
fix indenting

[ci skip]

CamJN committed:
add more virtual overrides for tests

CamJN committed:
ensure async request succeeds

CamJN committed:
make sure client disconnect wins race

CamJN committed:
comments explaining magic numbers

FooBarWidget committed:
fix indenting
[ci skip]

CamJN committed:
fix test 60

FooBarWidget committed:
only mock connect timeout in tests where we want to test the timeout

also make sessionSocketConnectTimeout() protected so we can properly override it

CamJN committed:
expand comment

[ci skip] [ci:skip]

CamJN committed:
drop test 62

FooBarWidget committed:
Make MAX_SESSION_CHECKOUT_TRY public

FooBarWidget committed:
Merge branch 'stable-6.0' into bugfix/nonblocking_socket

FooBarWidget commented on src/agent/Core/Controller.h:
```diff
@@ -355,9 +357,15 @@ class Controller: public ServerKit::HttpServer<Controller, Client> {
 
 	virtual void asyncGetFromApplicationPool(Request *req,
 		ApplicationPool2::GetCallback callback);
+	virtual int ioConditions();
```
You should use more specific naming, this name is too general and reveals too little context. Something like "getSessionSocketConnectIoWatchConditions", while verbose, is clearer.

You should also make it const.

I've also moved this method from the private section to the protected section, specifically to the subsection for mock overridable methods.

CamJN committed:
update docs to match new behaviour

[ci skip] [ci:skip]

FooBarWidget commented on src/agent/Core/Controller.h:
```diff
@@ -355,9 +357,15 @@ class Controller: public ServerKit::HttpServer<Controller, Client> {
 	virtual void asyncGetFromApplicationPool(Request *req,
 		ApplicationPool2::GetCallback callback);
+	virtual int ioConditions();
+	virtual double sessionSocketConnectTimeout(Request *req);
```
This name can be better. It sounds too much like a command, but it's actually a const getter. There may also be confusion between this and req->connectTimeout. So I think a better name would be "getSessionSocketEffectiveConnectTimeout".

And also make it const.

FooBarWidget commented on src/agent/Core/Controller.h:
```diff
@@ -355,9 +357,15 @@ class Controller: public ServerKit::HttpServer<Controller, Client> {
 public:
+	// If you change this value, make sure that Request::sessionCheckoutTry
+	// has enough bits.
+	static constexpr unsigned int MAX_SESSION_CHECKOUT_TRY = 10;
```
I moved this to public so you can reference it from tests.

FooBarWidget commented on src/agent/Core/ApplicationPool/TestSession.h:
```diff
@@ -162,13 +176,57 @@ class TestSession: public AbstractSession {
 		boost::lock_guard<boost::mutex> l(syncher);
 		return wantKeepAlive;
 	}
+	void setupAsync(){
```
This name is a bit vague and can be better. I propose "forceNonInstantConnect". Be sure to update the variable name too.

FooBarWidget commented on test/cxx/Core/ControllerTest.cpp:
```diff
@@ -53,21 +54,43 @@ namespace tut {
 			ApplicationPool2::AbstractSessionPtr mockSession;
 			ApplicationPool2::ExceptionPtr mockException;
+			unsigned int mockedSessionCount;
+			bool forceAsyncToWait;
```
I think this name can be better. If we consider the goal of this flag, how about "forceSessionSocketConnectTimeout".

CamJN replied:
I suppose, originally I had hoped to be able to allow the connect later (for additional test scenarios that would enable) which wouldn't fit with `forceSessionSocketConnectTimeout` but since that's not possible with this impl, I guess it fits.

FooBarWidget committed:
minor refactoring

[ci skip]

CamJN committed:
make test 61 thread safe

FooBarWidget committed:
Update test descriptions

[ci skip]

FooBarWidget commented:
We should have one final manual integration test, just to be on the safe side. @tinco  can help you with this.

1. Create an app that:
   1. Has a long-running request endpoint (e.g. `sleep 60`)
   2. The endpoint has a concurrency limit of 3, like a semaphore.
2. Run the app in Passenger Enterprise with thread count 4, max instances 1.
3. Make 5 simultaneous requests to the app.
4. Test that, while the 5 requests are in progress, that:
   1. On stable-6.0, `passenger-status --show=requests` freezes.
   2. On the PR branch, it does not freeze. It should show the 5th request being queued.
   3. All 5 requests eventually finish successfully. The first 3 requests should finish after 60 seconds, the 4th and 5th one after 120 seconds.

CamJN committed:
rename a bunch of things per review

FooBarWidget committed:
minor cleanups

CamJN commented:
That test process does not work, only 1 request is started at a time, not 3 and passenger-status does not freeze even on 6.0.26.

Here's the controller action I used:
```
class ApplicationController < ActionController::Base
  allow_browser versions: :modern
  @@semaphore = Concurrent::Semaphore.new(3)

  def slow
    @@semaphore.acquire
    sleep 60
    render plain: 'ok'
  end
end
```

This is the log after 60s (on 6.0.26):
```
[Thr-4] Started GET "/slow" for 127.0.0.1 at 2025-03-18 19:22:02 -0600
[Thr-4] Processing by ApplicationController#slow as */*
[Thr-4]   Parameters: {"method" => "slow"}
[Thr-4]   Rendering text template
[Thr-4]   Rendered text template (Duration: 0.3ms | GC: 0.0ms)
[Thr-4] Completed 200 OK in 60030ms (Views: 1.1ms | ActiveRecord: 0.0ms (0 queries, 0 cached) | GC: 0.0ms)


[Thr-1]   Rendering text template
[Thr-1]   Rendered text template (Duration: 1.1ms | GC: 0.0ms)
[Thr-1] Completed 200 OK in 60026ms (Views: 1.6ms | ActiveRecord: 0.0ms (0 queries, 0 cached) | GC: 0.0ms)


[ N 2025-03-18 19:22:13.8630 56307/Te Ser/Server.h:879 ]: [ServerThr.3] Freed 0 spare client objects
[ N 2025-03-18 19:22:13.8631 56307/Te Ser/Server.h:536 ]: [ServerThr.3] Shutdown finished
[ N 2025-03-18 19:22:13.8720 56307/Tg Ser/Server.h:879 ]: [ServerThr.4] Freed 0 spare client objects
[ N 2025-03-18 19:22:13.8720 56307/Tg Ser/Server.h:536 ]: [ServerThr.4] Shutdown finished
[Thr-4] Started GET "/slow" for 127.0.0.1 at 2025-03-18 19:22:13 -0600
[Thr-4] Processing by ApplicationController#slow as */*
[Thr-4]   Parameters: {"method" => "slow"}
```

And here's the log after a few minutes (on 6.0.26):
```
[Thr-4] Started GET "/slow" for 127.0.0.1 at 2025-03-18 19:22:02 -0600
[Thr-4] Processing by ApplicationController#slow as */*
[Thr-4]   Parameters: {"method" => "slow"}
[Thr-4]   Rendering text template
[Thr-4]   Rendered text template (Duration: 0.3ms | GC: 0.0ms)
[Thr-4] Completed 200 OK in 60030ms (Views: 1.1ms | ActiveRecord: 0.0ms (0 queries, 0 cached) | GC: 0.0ms)


[Thr-1]   Rendering text template
[Thr-1]   Rendered text template (Duration: 1.1ms | GC: 0.0ms)
[Thr-1] Completed 200 OK in 60026ms (Views: 1.6ms | ActiveRecord: 0.0ms (0 queries, 0 cached) | GC: 0.0ms)


[ N 2025-03-18 19:22:13.8630 56307/Te Ser/Server.h:879 ]: [ServerThr.3] Freed 0 spare client objects
[ N 2025-03-18 19:22:13.8631 56307/Te Ser/Server.h:536 ]: [ServerThr.3] Shutdown finished
[ N 2025-03-18 19:22:13.8720 56307/Tg Ser/Server.h:879 ]: [ServerThr.4] Freed 0 spare client objects
[ N 2025-03-18 19:22:13.8720 56307/Tg Ser/Server.h:536 ]: [ServerThr.4] Shutdown finished
[Thr-4] Started GET "/slow" for 127.0.0.1 at 2025-03-18 19:22:13 -0600
[Thr-4] Processing by ApplicationController#slow as */*
[Thr-4]   Parameters: {"method" => "slow"}
[ W 2025-03-18 19:22:20.4431 57599/T1 age/Wat/WatchdogMain.cpp:515 ]: Some Phusion Passenger(R) agent processes did not exit in time, forcefully shutting down all.
[Thr-4]   Rendering text template
[Thr-4]   Rendered text template (Duration: 0.1ms | GC: 0.0ms)
[Thr-4] Completed 200 OK in 60003ms (Views: 0.6ms | ActiveRecord: 0.0ms (0 queries, 0 cached) | GC: 0.0ms)


[Thr-4] Started GET "/slow" for 127.0.0.1 at 2025-03-18 19:23:02 -0600
[Thr-2] Started GET "/slow" for 127.0.0.1 at 2025-03-18 19:23:02 -0600
[Thr-3] Started GET "/slow" for 127.0.0.1 at 2025-03-18 19:23:02 -0600
[Thr-1] Started GET "/slow" for 127.0.0.1 at 2025-03-18 19:23:02 -0600
[Thr-3] Processing by ApplicationController#slow as */*
[Thr-3]   Parameters: {"method" => "slow"}
[Thr-4] Processing by ApplicationController#slow as */*
[Thr-4]   Parameters: {"method" => "slow"}
[Thr-1] Processing by ApplicationController#slow as */*
[Thr-1]   Parameters: {"method" => "slow"}
[Thr-2] Processing by ApplicationController#slow as */*
[Thr-2]   Parameters: {"method" => "slow"}
```

I think we should figure out how to test this on 6.0.26 first before testing 6.0.27.


I used this to send requests: `ab -s 1200 -c 5 -n 10 'http://127.0.0.1:3000/slow'`
Here's the `Passengerfile.json`:
```
{
  "thread_count": 4,
  "max_pool_size": 1,
  "concurrency_model": "thread"
}
```

FooBarWidget commented:
Ah you probably can't trigger the event loop getting blocked by a blocking connect. The socket backlog has to be full in order for that to happen. Ok how about this: either on the Ruby side you set a small backlog side, or you make a lot more simultaneous requests in order to fill the backlog. Also force concurrency to 0.

Also, your app never releases the semaphore? It should do that before the request ends.

Updated test plan:

1. App has a semaphore with limit 3, sleeps 30 seconds.
2. Run the app in Passenger Enterprise with thread count 4, max instances 1, force concurrency to 0.
3. Make 1000 simultaneous requests to the app.
4. Test that, while the 1000 requests are in progress, that:
   1. On stable-6.0, `passenger-status --show=requests` freezes or takes long to respond.
   2. On the PR branch, it does not freeze. It should show many requests being queued.
   3. All requests eventually finish successfully. If it takes too long you can lower the sleep time.

CamJN committed:
whitespace

[ci skip] [ci:skip]

CamJN commented:
On stable-6.0, `passenger-status --show=requests` does not freeze or take long to respond, with that test setup.

PassengerFile.json
```
{
  "thread_count": 4,
  "max_pool_size": 1,
  "force_max_concurrent_requests_per_process": 0,
  "concurrency_model": "thread"
}
```

command: `ab -s 1200 -c 1000 -n 1000 'http://127.0.0.1:3000/slow'`

App adjusted to 30s wait and releasing semaphore:
```
def slow
  @@semaphore.acquire
  sleep 30               
  render plain: 'ok'
  @@semaphore.release    
end
```

CamJN commented:
Testing 6.0.26 enterprise with: `wrk -c 1000 -t 1000 --timeout 1200 'http://127.0.0.1:3000/slow'`

`PassengerFile.json`:
```
{
  "max_pool_size": 1,
  "force_max_concurrent_requests_per_process": 0,
  "thread_count": 4,
  "concurrency_model": "thread"
}
```

Results in 0 requests in both top level and app-group queues.

`passenger-status` returns quickly with:
```
Version : 6.0.26
Date    : 2025-03-24 11:43:41 -0600
Instance: MnpiXejs (nginx/1.26.2 Phusion_Passenger/6.0.26)

----------- General information -----------
Max pool size : 1
App groups    : 1
Processes     : 1
Requests in top-level queue : 0

----------- Application groups -----------
/Users/camdennarzt/Developer/Ruby/testapp (development):
  App root: /Users/camdennarzt/Developer/Ruby/testapp
  Requests in queue: 0
  * PID: 3990    Sessions: 76      Processed: 0       Uptime: 26s
    CPU: 0%      Memory  : 71M     Last used: 26s ago
  * PID: 3922    Sessions: 129      Processed: 108     Uptime: 31s
    CPU: 0%      Memory  : 0M      Last used: 31s ago
    Shutting down...
  * PID: 3931    Sessions: 132      Processed: 107     Uptime: 30s
    CPU: 0%      Memory  : 0M      Last used: 30s ago
    Shutting down...
  * PID: 3940    Sessions: 132      Processed: 107     Uptime: 29s
    CPU: 0%      Memory  : 0M      Last used: 29s ago
    Shutting down...
  * PID: 3948    Sessions: 132      Processed: 108     Uptime: 29s
    CPU: 0%      Memory  : 0M      Last used: 29s ago
    Shutting down...
  * PID: 3957    Sessions: 132      Processed: 104     Uptime: 28s
    CPU: 0%      Memory  : 0M      Last used: 28s ago
    Shutting down...
  * PID: 3970    Sessions: 132      Processed: 89      Uptime: 27s
    CPU: 0%      Memory  : 0M      Last used: 27s ago
    Shutting down...
  * PID: 3979    Sessions: 132      Processed: 59      Uptime: 27s
    CPU: 0%      Memory  : 0M      Last used: 26s ago
    Shutting down...
```

`passenger-status --show=requests` returns quickly as well.

The client numbers in the logs range from [1-1] to [12-413].


With opensource 6.0.26:

`Passengerfile.json`:
```
{
  "max_pool_size": 1,
  "force_max_concurrent_requests_per_process": 0
}
```


The client numbers range from [1-1] to [12-954].
`passenger-status --show=requests` returns quickly.
`passenger-status` gives:
```
Version : 6.0.26
Date    : 2025-03-24 12:00:43 -0600
Instance: DQQ9RoqQ (nginx/1.26.2 Phusion_Passenger/6.0.26)

----------- General information -----------
Max pool size : 1
App groups    : 1
Processes     : 1
Requests in top-level queue : 0

----------- Application groups -----------
/Users/camdennarzt/Developer/Ruby/testapp (development):
  App root: /Users/camdennarzt/Developer/Ruby/testapp
  Requests in queue: 0
  * PID: 11970   Sessions: 92      Processed: 5       Uptime: 2m 53s
    CPU: 0%      Memory  : 76M     Last used: 2m 53s ago
  * PID: 11856   Sessions: 124      Processed: 109     Uptime: 2m 58s
    CPU: 0%      Memory  : 0M      Last used: 2m 58s ago
    Shutting down...
  * PID: 11865   Sessions: 124      Processed: 111     Uptime: 2m 57s
    CPU: 0%      Memory  : 0M      Last used: 2m 57s ago
    Shutting down...
  * PID: 11873   Sessions: 124      Processed: 110     Uptime: 2m 56s
    CPU: 0%      Memory  : 0M      Last used: 2m 56s ago
    Shutting down...
  * PID: 11882   Sessions: 124      Processed: 102     Uptime: 2m 56s
    CPU: 0%      Memory  : 0M      Last used: 2m 56s ago
    Shutting down...
  * PID: 11890   Sessions: 124      Processed: 103     Uptime: 2m 55s
    CPU: 0%      Memory  : 0M      Last used: 2m 55s ago
    Shutting down...
  * PID: 11914   Sessions: 124      Processed: 94      Uptime: 2m 55s
    CPU: 0%      Memory  : 0M      Last used: 2m 54s ago
    Shutting down...
  * PID: 11922   Sessions: 124      Processed: 78      Uptime: 2m 54s
    CPU: 0%      Memory  : 0M      Last used: 2m 54s ago
    Shutting down…
```

I noice that Passenger is ignoring the `"max_pool_size": 1`, in the `PassengerFile.json` in both OSS and Enterprise versions. Which reminds me of https://github.com/phusion/passenger/issues/2537.

CamJN committed:
Merge branch 'stable-6.0' into bugfix/nonblocking_socket

FooBarWidget committed:
Add trace points

[ci skip]

FooBarWidget commented:
I've tested and I've got good news, bad news and surprising news.
- Good news: the PR code does not break. It works.
- Bad and surprising news: I cannot reproduce a blocking connect. Neither on Linux nor on macOS. Things work differently than we expect.

On macOS, overflowing the socket backlog for a Unix domain socket, does not result in a blocking connect, but results in a "Connection refused" error!
On Linux, there seems to be no limit to the socket backlog, not even if the server requests a low backlog size. I cannot get the connect to block on Linux. There could still be cases where the connect block, but I cannot find it.

The "connection refused" error on macOS causes another surprise: it's the reason why we see more app processes than max pool size allows!
1. When Core Controller encounters "connection refused" when initiating a session, it assumes that the process is dead, so it detaches it.
2. That in turn spawns a new process.
3. The detached process isn't actually dead. It's still running the 'sleep' call. So it sticks around for a while.

What does this finding mean for us?
- We need to change the way we handle "Connection refused" errors when initiating the socket. Don't just assume the process is dead. This requires a separate discussion.
- I consider PR to be good enough. Camden and Tinco, please perform some general stress tests on the branch to check that it's stable. If your general tests pass then you can merge the PR and release.

FooBarWidget committed:
Improve error message

[ci skip]

CamJN committed:
Merge branch 'stable-6.0' into bugfix/nonblocking_socket

CamJN committed:
Merge branch 'stable-6.0' into bugfix/nonblocking_socket

CamJN committed:
debug plucky build
